keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2144 @@
+  when drain is requested by the maintenance object.
+- NodeMaintenance condition or annotation can be set on the node object to advertise the current
+  phase of the maintenance.
+
+#### Supported Stage Transitions
+
+The following transitions should be validated by the API server.
+
+- Idle -> _Deletion_
+  - Planning a maintenance in the future and canceling/deleting it without any consequence.
+- (Idle) -> Cordon -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable and then schedulable again.
+  - The complete stage will always be run even without specifying it.
+- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
+  - Cordon and Complete stages will always be run, even without specifying them.
+- (Idle) -> Complete -> _Deletion_.
+  -  Make a set of nodes schedulable.
+
+The stage transitions are invoked either manually by the cluster admin or by a higher-level
+controller. For a simple drain, cluster admin can simply create the NodeMaintenance with
+`stage: Drain` directly.
+
+### DaemonSet Controller
+
+The DaemonSet workloads should be tied to the node lifecycle because they typically run critical
+workloads where availability is paramount. Therefore, the daemonset controller should respond to
+the Evacuation only if there is a NodeMaintenance happening on that node and the DaemonSet is in
+the `drainTargets`. For example, if we observe the following NodeMaintenance:
+
+```yaml
+apiVersion: v1alpha1
+kind: NodeMaintenance
+...
+status:
+  nodeStatuses:
+    - nodeRef:
+        name: six
+      drainTargets:
+        - podPriority: 2147483647
+          podType: Default
+        - podPriority: 5000
+          podType: DaemonSet
+  ...
+```
+
+To register the daemonset controller as a controller evacuator, it should set the following
+annotation on its own pods.
+
+```yaml
+evacuation.coordination.k8s.io/daemonset.apps.k8s.io: "10000/controller"
+```
+
+The controller should respond to the Evacuation object when it observes its own class
+(`daemonset.apps.k8s.io`) in `.status.activeEvacuatorClass`.
+
+For the above node maintenance, the controller should not react to Evacuations of DaemonSet pods
+with a priority greater than 5000. This state should not normally occur, as Evacuation requests
+should be coordinated with NodeMaintenance. If it does occur, we should not encourage this flow by
+updating the `.status.ActiveEvacuatorCompleted` field, although it is required to update this field
+for normal workloads.
+
+If the DaemonSet pods have a priority equal to or less than 5000, the Evacuation status should be
+updated appropriately as follows, and the targeted pod should be deleted by the daemonset
+controller:
+
+```yaml
+apiVersion: v1alpha1
+kind: Evacuation
+metadata:
+  finalizers:
+    evacuation.coordination.k8s.io/instigator_nodemaintenance.k8s.io
```

Comment:
Do we typically use `_` in finalizer names? It doesn't quite fit what I'd expect.

Reply:
Please see https://github.com/kubernetes/enhancements/pull/4213#discussion_r1600458460 for an explanation.

FYI, we use finalizers:
- To ensure that only the instigator can cancel the evacuation and that Evacuations are not stomped over/deleted by other parties.
- To track multiple instigators.
- To allow reaction to deletions by the instigator and the evacuator. Most likely they will want to atomically complete the evacuation. 
- It is in this form for user convenience so we can automatically remove Evacuations when the pod is removed.

for additional information, please see
- https://github.com/kubernetes/enhancements/blob/2d7dbab6c95575d68395c30fa789f87d9a2c3c01/keps/sig-apps/4563-evacuation-api/README.md#evacuation-instigator-1
- https://github.com/kubernetes/enhancements/blob/2d7dbab6c95575d68395c30fa789f87d9a2c3c01/keps/sig-apps/4563-evacuation-api/README.md#evacuation-instigator-finalizer